lib/matplotlib/layoutbox.py

```diff
@@ -0,0 +1,727 @@
+# -*- coding: utf-8 -*-
+"""
+
+Conventions:
+
+"constrain_x" means to constrain the variable with either
+another kiwisolver variable, or a float.  i.e. `constrain_width(0.2)`
+will set a constraint that the width has to be 0.2 and this constraint is
+permanent - i.e. it will not be removed if it becomes obsolete.
+
+"edit_x" means to set x to a value (just a float), and that this value can
+change.  So `edit_width(0.2)` will set width to be 0.2, but `edit_width(0.3)`
+will allow it to change to 0.3 later.  Note that these values are still just
+"suggestions" in `kiwisolver` parlance, and could be over-ridden by
+other constrains.
+
+"""
+
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals)
+import kiwisolver as kiwi
+import numpy as np
+import matplotlib
+import warnings
+
+
+# renderers can be complicated
+def get_renderer(fig):
+    if fig._cachedRenderer:
+        renderer = fig._cachedRenderer
+    else:
+        canvas = fig.canvas
+        if canvas and hasattr(canvas, "get_renderer"):
+            renderer = canvas.get_renderer()
+        else:
+            # not sure if this can happen
+            # seems to with PDF...
+            #warnings.warn("constrained_layout : falling back to Agg renderer")
+            from matplotlib.backends.backend_agg import FigureCanvasAgg
+            canvas = FigureCanvasAgg(fig)
+            renderer = canvas.get_renderer()
+
+    return renderer
+
+
+class LayoutBox(object):
+    """
+    Basic rectangle representation using kiwi solver variables
+    """
+
+    def __init__(self, parent=None, name='', tightwidth=False,
+                 tightheight=False, artist=None,
+                 lower_left=(0, 0), upper_right=(1, 1), pos=False,
+                 subplot=False, h_pad=None, w_pad=None):
+        Variable = kiwi.Variable
```

Comment:
Sorry to be a bit anal with this but again
1. does the module need to be public?
2. do all attributes of this class need to be public?
(I'm a bit scar(r)ed with big public APIs...)

Reply:
I suppose it should be private.  @has2k1  had a similar suggestion.  Hmmm, now I have to figure out how to make a whole module private 🐑